src/transformers/models/timesformer/modeling_timesformer.py

```diff
@@ -0,0 +1,856 @@
+    random_tensor = keep_prob + torch.rand(shape, dtype=input.dtype, device=input.device)
+    random_tensor.floor_()  # binarize
+    output = input.div(keep_prob) * random_tensor
+    return output
+
+
+class TimeSformerDropPath(nn.Module):
+    """Drop paths (Stochastic Depth) per sample (when applied in main path of residual blocks)."""
+
+    def __init__(self, config: TimeSformerConfig):
+        super().__init__()
+        self.drop_prob = config.drop_path_prob
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return drop_path(x, self.drop_prob, self.training)
+
+    def extra_repr(self) -> str:
+        return "p={}".format(self.drop_prob)
+
+
+# Adapted from https://github.com/facebookresearch/TimeSformer/blob/a5ef29a7b7264baff199a30b3306ac27de901133/timesformer/models/vit.py#L57
+# and transformers.models.videomae.modeling_videomae.VideoMAESelfAttention
+class TimeSformerSelfAttention(nn.Module):
+    def __init__(self, config: TimeSformerConfig):
+        super().__init__()
+
+        dim = config.hidden_size
+        num_heads = config.num_attention_heads
+        qkv_bias = config.qkv_bias
+        attn_drop = config.attention_probs_dropout_prob
+
+        self.num_heads = num_heads
+        head_dim = dim // num_heads
+        self.scale = head_dim**-0.5
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+
+    def forward(self, hidden_states, output_attentions: bool = False):
+        B, N, C = hidden_states.shape
+        qkv = self.qkv(hidden_states).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
+        q, k, v = qkv[0], qkv[1], qkv[2]
+
+        attention_probs = (q @ k.transpose(-2, -1)) * self.scale
+        attention_probs = attention_probs.softmax(dim=-1)
+        attention_probs = self.attn_drop(attention_probs)
+
+        context_layer = (attention_probs @ v).transpose(1, 2).reshape(B, N, C)
+
+        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+
+        return outputs
+
+
+# copied from transformers.models.videomae.modeling_videomae.VideoMAESelfOutput
+class TimeSformerSelfOutput(nn.Module):
+    """
+    The residual connection is defined in TimeSformerLayer instead of here (as is the case with other models), due to
+    the layernorm applied before each block.
+    """
+
+    def __init__(self, config: TimeSformerConfig) -> None:
+        super().__init__()
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+
+        return hidden_states
+
+
+# copied from transformers.models.videomae.modeling_videomae.VideoMAEAttention
```

Comment:
```suggestion
# Copied from transformers.models.videomae.modeling_videomae.VideoMAEAttention with VideoMAE->TimeSformer
```

Reply:
it seems the function has some differences, so I changed `Copied ` with `Adapted`.